pydantic_ai_graph/pydantic_ai_graph/graph.py

```
@@ -2,169 +2,232 @@
-# noinspection PyTypeHints
 @dataclass(init=False)
-class Graph(Generic[StateT, GraphInputT, GraphOutputT]):
+class Graph(Generic[StateT, RunEndT]):
     """Definition of a graph."""
 
-    first_node: NodeDef[StateT, Any, Any]
-    nodes: dict[str, NodeDef[StateT, Any, Any]]
-    name: str | None
+    name: str
+    nodes: tuple[type[BaseNode[StateT, RunEndT]], ...]
+    node_defs: dict[str, NodeDef[StateT, RunEndT]]
 
     def __init__(
         self,
-        first_node: type[BaseNode[StateT, GraphInputT, GraphOutputT]],
-        *other_nodes: type[BaseNode[StateT, Any, GraphOutputT]],
-        name: str | None = None,
+        name: str = 'graph',
+        nodes: tuple[type[BaseNode[StateT, RunEndT]], ...] = (),
         state_type: type[StateT] | None = None,
     ):
+        self.name = name
+
+        _nodes_by_id: dict[str, type[BaseNode[StateT, RunEndT]]] = {}
+        for node in nodes:
+            node_id = node.get_id()
+            if (existing_node := _nodes_by_id.get(node_id)) and existing_node is not node:
+                raise ValueError(f'Node ID "{node_id}" is not unique — found in {existing_node} and {node}')
+            else:
+                _nodes_by_id[node_id] = node
+        self.nodes = tuple(_nodes_by_id.values())
+
         parent_namespace = get_parent_namespace(inspect.currentframe())
-        self.first_node = first_node.get_node_def(parent_namespace)
-        self.nodes = nodes = {self.first_node.node_id: self.first_node}
-        for node in other_nodes:
-            node_def = node.get_node_def(parent_namespace)
-            nodes[node_def.node_id] = node_def
+        _node_defs: dict[str, NodeDef[StateT, RunEndT]] = {}
+        for node in self.nodes:
+            _node_defs[node.get_id()] = node.get_node_def(parent_namespace)
+        self.node_defs = _node_defs
 
-        self._check()
-        self.name = name
+        self._validate_edges()
+
+    def validate_next_node(self, next_node: BaseNode[StateT, RunEndT]) -> None:
```

this isn't used, is it necessary? If so, what's it for?

---

I was using it, then broke it, but basically I wanted to set things up so that during a run it would check dynamically that the node was in the graph. But I guess I'm not sure if it's better to raise an error if you return something not in the expected graph or just chug along unimpacted given it doesn't _need_ to error. Maybe a warning is a good compromise?

---

Specifically, I was using it in a callback, but then removed the callbacks stuff I had added since I think there's a decent amount of discussion to be had about the exact implementation details

---

I removed this method for now